jpa/org.eclipse.persistence.jpa/src/main/java/org/eclipse/persistence/jpa/metadata/FileBasedProjectCacheFilter.java

```diff
@@ -0,0 +1,128 @@
+            // Check current object instance.
+            final Class<?> serialClass = info.serialClass();
+            if (serialClass != null) {
+                log.log(SessionLog.FINE, String.format(
+                        "FileBasedProjectCacheFilter: processing level %d class %s",
+                        info.depth(), serialClass.getName()));
+                // Only Project class instance is allowed on top level of the structure
+                // and class must also come from known .jar file name.
+                if (info.depth() == 1L) {
+                    if (serialClass != Project.class || !isEclipseLinkFileJar(serialClass, log)) {
+                        log.log(SessionLog.WARNING, String.format(
+                                "FileBasedProjectCacheFilter: rejected illegal top level FileBasedProjectCache class: %s",
+                                serialClass.getName()));
+                        return Status.REJECTED;
+                    }
+                }
+                // Remote interface must be rejected
+                if (Remote.class.isAssignableFrom(serialClass)) {
+                    log.log(SessionLog.WARNING, String.format(
+                            "FileBasedProjectCacheFilter: rejected illegal Remote FileBasedProjectCache class: %s",
+                            serialClass.getName()));
+                    return Status.REJECTED;
+                }
+            }
+            return Status.UNDECIDED;
+        } catch (Throwable t) {
+            log.log(SessionLog.WARNING, "Exception in FileBasedProjectCacheFilter check", t);
+            throw t;
+        }
+    }
+
+    // Check whether Eclipselink own class is from known .jar file name.
+    private static boolean isEclipseLinkFileJar(final Class<?> serialClass, final SessionLog log) {
+        final URL classUrl = serialClass.getResource(
+                '/' + serialClass.getName().replace('.', '/') + ".class");
+        if (classUrl != null && "jar".equals(classUrl.getProtocol())) {
+            final String classFile = classUrl.getPath();
+            try {
+                final URL fileUrl = new URL(classFile);
+                if ("file".equals(fileUrl.getProtocol())) {
+                    final String filePaths = fileUrl.getPath();
+                    final int delimPos = filePaths.indexOf('!');
+                    String file;
+                    if (delimPos > 0) {
+                        file = filePaths.substring(0, delimPos);
+                    } else {
+                        file = filePaths;
+                    }
+                    final int pathSepPos = file.lastIndexOf('/');
+                    if (pathSepPos > 0 && file.length() > pathSepPos + 1) {
+                        file = file.substring(pathSepPos + 1);
+                    }
+                    if ((file.startsWith("org.eclipse.persistence") && file.contains(Version.getVersion()))
+                            || file.equals("eclipselink.jar")) {
```

Comment:
Well, it would be nice to have some list of all file names that can be used.

Reply:
I removed whole file names check. Unfortunately we can't rely on known set of file names because of various customer patches, file name changes in application servers, etc.